File: samples/snippets/csharp/VS_Snippets_CLR/Type_GetInterface/CS/type_getinterface.cs
﻿/*
   System.Type.GetInterface(String)
   System.Type.GetInterface(String, bool)
   System.Type.GetInterfaceMap

   The following program get the type of Hashtable class and searches for the interface
   with the specified name. Then prints the method name of that interface.
*/

using System;
using System.Reflection;
using System.Collections;

class MyInterfaceClass
{
// <Snippet1>
// <Snippet2>
// <Snippet3>
    public static void Main()
    {
        Hashtable hashtableObj = new Hashtable();
        Type objType = hashtableObj.GetType();
        MethodInfo[] arrayMethodInfo;
        MemberInfo[] arrayMemberInfo;
        try
        {
            // Get the methods implemented in 'IDeserializationCallback' interface.
            arrayMethodInfo =objType.GetInterface("IDeserializationCallback").GetMethods();
            Console.WriteLine ("\nMethods of 'IDeserializationCallback' Interface :");
            foreach(MethodInfo methodInfo in arrayMethodInfo)
                Console.WriteLine (methodInfo);

            // Get FullName for interface by using Ignore case search.
            Console.WriteLine ("\nMethods of 'IEnumerable' Interface");
            arrayMethodInfo = objType.GetInterface("ienumerable",true).GetMethods();
            foreach(MethodInfo methodInfo in arrayMethodInfo)
               Console.WriteLine (methodInfo);

            //Get the Interface methods for 'IDictionary' interface
            InterfaceMapping interfaceMappingOb = objType.GetInterfaceMap(typeof(IDictionary));
            arrayMemberInfo = interfaceMappingObj.InterfaceMethods;
            Console.WriteLine ("\nHashtable class Implements the following IDictionary Interface methods :");
            foreach(MemberInfo memberInfo in arrayMemberInfo)
               Console.WriteLine (memberInfo);
        }
        catch (Exception e)
        {
            Console.WriteLine ("Exception : " + e.ToString());
        }
    }
// </Snippet3>
// </Snippet2>
// </Snippet1>
}

File: snippets/fsharp/System/Type/GetInterface/type_getinterface.fs
open System.Collections

// <Snippet1>
// <Snippet2>
// <Snippet3>
let hashtableObj = Hashtable()
let objType = hashtableObj.GetType()
try
    // Get the methods implemented in 'IDeserializationCallback' interface.
    let arrayMethodInfo = objType.GetInterface("IDeserializationCallback").GetMethods()
    printfn "\nMethods of 'IDeserializationCallback' Interface :"
    for methodInfo in arrayMethodInfo do
        printfn $"{methodInfo}"

    // Get FullName for interface by using Ignore case search.
    printfn "\nMethods of 'IEnumerable' Interface"
    let arrayMethodInfo = objType.GetInterface("ienumerable",true).GetMethods()
    for methodInfo in arrayMethodInfo do
        printfn $"{methodInfo}"

    //Get the Interface methods for 'IDictionary' interface
    let interfaceMappingObj = objType.GetInterfaceMap typeof<IDictionary>
    let arrayMemberInfo = interfaceMappingObj.InterfaceMethods
    printfn "\nHashtable class Implements the following IDictionary Interface methods :"
    for memberInfo in arrayMemberInfo do
        printfn $"{memberInfo}"
with e ->
    printfn $"Exception : {e}"
// </Snippet3>
// </Snippet2>
// </Snippet1>